### DISCUSSION TITLE
qpdf 10.6.3.0cmake1

### DISCUSSION THREAD

**Discussion by jberkenbilt** [qpdf/qpdf#676] [General]
This is qpdf version 10.6.3.0cmake1. It is identical in functionality and binary compatible with 10.6.3. The only changes are
* The old autoconf + home-grown Makefile-based build system has been replaced with cmake
* For Windows, releases contain additional files (examples, static libraries) and executable installers are available in addition to zip files
* The documentation has been substantially updated to describe the new build system. In particular, see
  * [Building and Installing QPDF](https://qpdf.readthedocs.io/en/release-qpdf-10.6.3.0cmake1/installation.html)
  * [Notes for Packagers](https://qpdf.readthedocs.io/en/release-qpdf-10.6.3.0cmake1/packaging.html)
  * [Converting From autoconf to cmake](https://qpdf.readthedocs.io/en/release-qpdf-10.6.3.0cmake1/installation.html#converting-from-autoconf-to-cmake)

This is a pre-release. Who should test it out?
* There are no functionality changes, so if you're just using qpdf, there's no need to take this release.
* If you packager qpdf for a distribution or build it yourself as part of a larger system, you can use this to migrate to cmake.
* If you are a Windows developer who has been having problems using the DLL from another language, you can try with this release and see if it helps. It's possible that the way cmake builds may resolve some issues that have so far remained elusive.

For a full list of changes from previous releases, please see the [release notes](https://qpdf.readthedocs.io/en/release-qpdf-10.6.3.0cmake1/release-notes.html). See also [README-what-to-download](./README-what-to-download.md) for details about the available source and binary distributions.

<hr /><em>This discussion was created from the release <a href='https://github.com/qpdf/qpdf/releases/tag/release-qpdf-10.6.3.0cmake1'>qpdf 10.6.3.0cmake1</a>.</em>

**Comment by jberkenbilt** [comment]
Packagers: here's the commit where I changed the debian package to use cmake. It's pretty simple as you can see. https://github.com/jberkenbilt/debian-qpdf/commit/4cc4e6f327972dc157707ee646c2e6a940fd6c79

Also, this is where I added autopkgtest. I don't know what systems non-debian-based distributions might have. The debian test files are one-liners that invoke the test files in the source distribution passing the temporary directory as an argument. https://github.com/jberkenbilt/debian-qpdf/commit/e927404b4409ce6653d0feb88bc1cca649c198d5

**Comment by jberkenbilt** [comment]
My configure-time check for whether -latomic was needed was inadequate as some platforms need it for `std::atomic<unsigned long long>` but not `std::atomic<int>`. There was also a scope error in CMakeLists.txt.  A fix is in main.

**Comment by thesamesam** [comment]
I've just done a test-run packaging this for Gentoo (not distributed to users) and everything went smoothly first time. Thanks for the excellent documentation on the migration! (I have to say that qpdf's packaging docs have always been very helpful anyway though!)

Thanks for the effort you're putting in to make this an easy transition.

**Reply by jberkenbilt** [reply]
Glad to hear it. Thanks!

**Comment by fredemmott** [comment]
Sorry for it taking so long for me to get to this - seems to be working fine :)

I'm doing static linking in a CMake project, where I also build zlib and libjpegturbo; I've specified the header and library paths explicitly instead of using the 'external-libraries' approach: https://github.com/fredemmott/OpenKneeboard/blob/2b02adf7fcf5da0e0564090dcd6b2c7160c38073/third-party/qpdf.cmake

**Reply by fredemmott** [reply]
Just to bring this back to the original issue: can confirm this works fine with all reasonable combinations of the MSVC C++ runtime:
- debug and release
- static, shared, and [MS's new 'Hybrid CRT' approach](https://github.com/microsoft/WindowsAppSDK/blob/main/docs/Coding-Guidelines/HybridCRT.md) - in short, dynamically linking the ABI-stable part of the runtime (UCRT) but statically linking the much smaller part that changes frequently (vcruntime/msvcp and friends)

**Reply by jberkenbilt** [reply]
Thanks for testing and reporting back.

**Comment by jberkenbilt** [comment]
I'm locking this discussion now since qpdf 11 (which uses cmake) is out.
